### PR TITLE
Don't create an AuthProvider if one exists

### LINKED ISSUE #624: Generators: what happens when an auth generator is run more than once?
Right now if an auth generator is run multiple times it will add more and more lines of config to `web/src/index.js` which isn't the greatest experience.

* Should the generator exit with an error if the first insert already exists? ~Or should it just skip it (with a logged note) and keep going? My first thought is to just skip the step if a quick `match()` says that the line that's going to be inserted is already there, but keep going to the next step.~ **Yep, the first thing we should check (before the first step of adding packages) is if `<AuthProvider>` already exists and if so, error out with a message.**
* ~What happens if you run two different auth generators, one after the other? My first thought is to check if `<AuthProvider>` already exists and if the `type` doesn't match the generator that's about to run, exit/error and leave a note that the other auth type has already been added.~ **This will be covered by the first solution.**

Opinions (and PRs) welcome!

### PR DESCRIPTION
**Pull request by amorriscode**
This PR:
- [x] Resolves #624 
- [x] Adds two basic auth generator tests

Result of command:

```bash
❯ yarn rw g auth netlify
yarn run v1.22.4
$ /Users/amorriscode/dev/example-blog/node_modules/.bin/rw g auth netlify
Existing auth provider found
✨  Done in 1.90s.
```

### CODE REVIEW

**Review by thedavidprice — CHANGES_REQUESTED** [review]
@amorriscode Thanks for getting this one going! We do indeed want to check for existing and default to _not_ overwriting. However, we also need to provide an option to override with the `--force` flag.

Take a look at how the `setup tailwind` command works here --> https://github.com/redwoodjs/redwood/blob/main/packages/cli/src/commands/setup/tailwind/tailwind.js

For each step in the process where a file is created or modified, there's a check if it exists. If so, it declines to run the step _unless_ `force` is passed. Example here:
https://github.com/redwoodjs/redwood/blob/66e39e67d11e44e789390aed4cae735689e5464e/packages/cli/src/commands/setup/tailwind/tailwind.js#L85-L96

Note: you don't really need to run this check for installation of packages because Yarn will handle that by default.

Does all this make sense?

**Comment by thedavidprice** [comment]
@amorriscode one more FYI. A next step here after this is merged could be to move the `auth.js` command (and provider directory) to be a part of the new Setup command. Each auth provider will be a specific setup command, e.g. `yarn rw setup auth0` or `yarn rw setup netlify-auth`.

Take a look at how I deprecated the previous `rw g util tailwind` command here https://github.com/redwoodjs/redwood/blob/main/packages/cli/src/commands/generate/util/tailwind and moved it here https://github.com/redwoodjs/redwood/tree/main/packages/cli/src/commands/setup/tailwind

If you do this, make sure we:
- give deprecation info in the files as well as the CLI output
- update the CLI Docs on the website (can just create an issue)
- create auth-specific names for each provider setup command (because in addition to `setup netlify-auth` we'll also eventually have `setup netlify-deploy`)

Having written this out, it might be a heavier lift than I initially thought. But just let me know either way. NO pressure!

**Comment by amorriscode** [comment]
@thedavidprice would love some feedback on the handling of `--force`. The file override was already in place in the handler for the auth lib file: https://github.com/redwoodjs/redwood/blob/66e39e67d11e44e789390aed4cae735689e5464e/packages/cli/src/commands/generate/auth/auth.js#L153-L162

To handle `--force` elsewhere, I'm detecting the current `AuthProvider`, replacing its `init` code, and updating the `AuthProvider` component. Let me know what you think! It seems to work alright on my end.

image

As far as the other task goes (moving `auth.js`), I think I'll tackle that in another PR!

**Review by thedavidprice — CHANGES_REQUESTED** [review]
Anthony! This is looking great and I can tell you're thinking through all the changes and edge cases really well. Lot of work here.

In addition to the change request inline, there's one more thing that's not working correctly on `--force`

The section starting L225 `title: 'Adding auth config to GraphQL API...',` doesn't have any conditional logic in case of `--force`. So what happens is that you end up with a double import:
```js
// api/src/functions/graphql.js
...
import schemas from 'src/graphql/**/*.{js,ts}'
import services from 'src/services/**/*.{js,ts}'

import { getCurrentUser } from 'src/lib/auth'

import { getCurrentUser } from 'src/lib/auth.js'
import { db } from 'src/lib/db'
...
```

So in case of `force`, you might just skip this step. A better way (more aligned with the work you've been doing here) would be to check if `import { getCurrentUser } from 'src/lib/auth'` exists and act accordingly.

Lastly, I did make several code change requests based on Prettier warnings. Mostly curious as to why my VS Code settings were flagging these and why they didn't just update on save on your end. Maybe it's me?

**Comment by amorriscode** [comment]
@thedavidprice can't believe I missed the GraphQL import! 😅 With the formatting, I feel like maybe my VS Code is just unhappy with so many plugins. This morning I realized there was a TypeScript/JS one that was conflicting with Prettier. I don't see any Prettier warnings on my code but at least when I run the formatter it seems to be happier now.

Might try starting fresh and seeing what happens.

**Comment by thedavidprice** [comment]
All good! I'll spend some time with this on Monday in hopes of getting it merged -- but it looks great first glance. Thank you so much.

File: packages/cli/src/commands/generate/auth/auth.js [hunk 1]
```diff
@@ -80,30 +124,42 @@ export const files = (provider) => {
...
 
 export const addApiConfig = () => {
   let content = fs.readFileSync(API_GRAPHQL_PATH).toString()
-
-  // add import statement
-  content = content.replace(
-    /^(.*services.*)$/m,
-    `$1\n\nimport { getCurrentUser } from 'src/lib/auth'`
+  const [_, hasAuthImport] = content.match(
+    /(import {.*} from 'src\/lib\/auth.*')/s
```
Comment (thedavidprice) [inline packages/cli/src/commands/generate/auth/auth.js:147]: This is throwing an error:
```bash
object null is not iterable (cannot read property Symbol(Symbol.iterator))
```

But I manually checked logic for case true | false and it's working well.

**Comment by thedavidprice** [comment]
Ok, _so close_ — just two things I found. First is the code comment about an error. Secondly, I didn't notice this before but check out these two templates files:
- Firebase: https://github.com/redwoodjs/redwood/blob/main/packages/cli/src/commands/generate/auth/providers/firebase.js
- magicLink: https://github.com/redwoodjs/redwood/blob/main/packages/cli/src/commands/generate/auth/providers/magicLink.js

...and compare them to...

- https://github.com/redwoodjs/redwood/blob/main/packages/cli/src/commands/generate/auth/providers/auth0.js

It turns out the first two files are missing the three comment lines, which are needed for the `--force` to work correctly on the file `index.js`. E.g. currently if you try to re-generate auth with `-f` and `firebase`, it seems you end up with the duplicate lines in `web/index.js`. But for `netlify` or `auth0` it finds and replaces very nicely.

So want to confirm you're seeing the same behavior I'm seeing here. And, if so, it seems the easy resolution is to:
- [ ] Add the three comment lines to `auth/providers/firebase.js` and `auth/providers/magicLink.js`

Sound good?

**Comment by amorriscode** [comment]
@thedavidprice I'm having trouble reproducing the issue. 😢 

Here is a screenshot after running `yarn rw g auth magicLink --force` on the example blog:
image

Here is a screenshot after running `yarn rw g auth firebase --force` on the example blog:
image

Two things to note:
1. There is some extra whitespace I'm having trouble getting rid of that you can see in both screenshots
2. I've found that I have to relink using Redwood tools after using `--force`; the last step of the auth gen installs packages which wipes out my link meaning if I run `auth --force` commands back to back, without relinking with `rwt`, it doesn't work

Sorry I haven't been able to resolve this on my own.

File: packages/cli/src/commands/generate/auth/auth.js (hunk 1 above)
Comment (thedavidprice) [reply packages/cli/src/commands/generate/auth/auth.js:147]: for `.match()` "Matches a string or an object that supports being matched against, and returns an array containing the results of that search, or null if no matches are found."

So I believe what's happening in the case where project does **not** already have auth, it returns `null` and throws.

**Comment by thedavidprice** [comment]
>There is some extra whitespace I'm having trouble getting rid of that you can see in both screenshots

^^ Ah, that's low priority.

>I've found that I have to relink using Redwood tools after using --force; the last step of the auth gen installs packages which wipes out my link meaning if I run auth --force commands back to back, without relinking with rwt, it doesn't work

^^ huh, that's also weird + frustrating. I've been reverting commits in my test project after each test, so maybe that's what's keeping me from having to stop, `yarn --force`, then `yarn rwt cpw ...` again.

> Sorry I haven't been able to resolve this on my own.

^^ Apology not needed. I'm here to help. And you're the one doing the work! Don't stop now.

**Comment by thedavidprice** [comment]
Ok, see my new comment inline regarding the error at line 146. Once you get setup, I think you'll hammer through this to the finish line 🏁

Here's my setup and recommendations:
1. fresh installation: `yarn create redwood-app test`
2. Git init, add, and commit
3. do the setup dance: in Framework `yarn build:watch`, and in test/ project `yarn rwt cpw ../dir/to-framework`
4. **Fix Error:** try running `yarn rw g auth netlify`, which should throw.
   - do magic to resolve
5. `yarn rw g auth netlify` On success, commit.
6. Open `web/src/index.js`
7.  Run `yarn rw g auth firebase -f` --> it'll have duplications in web/src/index.js, most noticeably `<AuthProvider ...`
8. Revert file changes
9. Add those three comment lines to the firebase and magicLink files
10. Run again with -f and it works correctly with no duplications ¯\_(ツ)_/¯
11. Ship. It.

**Comment by amorriscode** [comment]
Thanks for the help @thedavidprice. I still think the issue with running the command twice is that the first `yarn rw g auth netlify` replaces the CLI that is linked with `yarn rwt cpw`. I could be wrong but try this...

## Test 1 (Fails)

1. Fresh install `yarn create redwood-app test`
2. Initial commit
3. In the framework `yarn build:watch` then test project `yarn rwt cpw ../dir/to-framework`
4. `yarn rw g auth netlify` (installs npm modules) --> commit
5. `yarn rw g auth firebase -f` --> bad force
6. Revert file changes

## Test 2 (Succeeds)

1. Fresh install `yarn create redwood-app test`
2. Initial commit
3. In the framework `yarn build:watch` then test project `yarn rwt cpw ../dir/to-framework`
4. `yarn rw g auth netlify` (installs npm modules) --> commit
5. **Restart `yarn rwt cpw ../dir/to-framwork` in the test project**
6. `yarn rw g auth firebase -f` --> everything looks good!

**Comment by thedavidprice** [comment]
>  I still think the issue with running the command twice is that the first yarn rw g auth netlify replaces the CLI that is linked with yarn rwt cpw

Gah, now my brain hurts! I could have sworn I check this the first go around and, therefore, assumed it was something with the regex matching (admittedly I didn't dig in to that part of the code). Makes sense that `yarn` running would bust the linking but then why... never mind. Let's get this thing merged.

Thank you and well done! 🚀

**Review by thedavidprice — APPROVED** [review]

**Pull request merged**